front/pages/api/w/[wId]/data_sources/[name]/databases/csv.ts

```diff
@@ -0,0 +1,324 @@
+          "Failed to create database."
+        );
+        return apiError(req, res, {
+          status_code: 500,
+          api_error: {
+            type: "internal_server_error",
+            message: "Failed to create database.",
+          },
+        });
+      }
+
+      const { database } = dbRes.value;
+
+      const tableId = generateModelSId();
+      const tableRes = await CoreAPI.upsertDatabaseTable({
+        projectId: dataSource.dustAPIProjectId,
+        dataSourceName: dataSource.name,
+        databaseId: id,
+        description,
+        name: name,
+        tableId,
+      });
+
+      if (tableRes.isErr()) {
+        logger.error(
+          {
+            dataSourceName: dataSource.name,
+            workspaceId: owner.id,
+            databaseName: name,
+            databaseId: id,
+            tableId,
+            tableName: name,
+            error: tableRes.error,
+          },
+          "Failed to upsert database table."
+        );
+
+        return apiError(req, res, {
+          status_code: 500,
+          api_error: {
+            type: "internal_server_error",
+            message: "Failed to upsert database table.",
+          },
+        });
+      }
+
+      const rowsRes = await CoreAPI.upsertDatabaseRows({
+        projectId: dataSource.dustAPIProjectId,
+        dataSourceName: dataSource.name,
+        databaseId: id,
+        tableId,
+        rows: csvRows,
+      });
+
+      if (rowsRes.isErr()) {
+        logger.error(
+          {
+            dataSourceName: dataSource.name,
+            workspaceId: owner.id,
+            databaseName: name,
+            databaseId: id,
+            tableId,
+            tableName: name,
+            error: rowsRes.error,
+          },
+          "Failed to upsert database rows."
+        );
+
+        return apiError(req, res, {
+          status_code: 500,
+          api_error: {
+            type: "internal_server_error",
+            message: "Failed to upsert database rows.",
+          },
+        });
+      }
+
+      return res.status(200).json({ database });
+
+    default:
+      return apiError(req, res, {
+        status_code: 405,
+        api_error: {
+          type: "method_not_supported_error",
+          message: "The method passed is not supported, POST is expected.",
+        },
+      });
+  }
+}
+
+export default withLogging(handler);
+
+async function rowsFromCsv(
+  csv: string
+): Promise<Result<CoreAPIDatabaseRow[], APIError>> {
+  const parser = parse(csv);
+  let header: string[] | undefined = undefined;
+  const valuesByCol: Record<string, string[]> = {};
+
+  for await (const anyRecord of parser) {
+    // Assert that record is string[].
+    if (!Array.isArray(anyRecord)) {
+      throw new Error("Record is not an array");
+    }
+    if (anyRecord.some((r) => typeof r !== "string")) {
+      throw new Error("Record contains non-string values");
+    }
+
+    const record = anyRecord as string[];
+
+    if (!header) {
```

Comment:
That means we never have properly named headers?

Reply:
Or is there an implicit assumption that the first record is a header? But then we're inserting it no?

Reply:
> Or is there an implicit assumption that the first record is a header? But then we're inserting it no?

Yes that's the assumption, and no we're not inserting it https://github.com/dust-tt/dust/blob/baf031991b974eee66f532cef3333718ba87ff85/front/pages/api/w/%5BwId%5D/data_sources/%5Bname%5D/databases/csv.ts#L213-L216

Reply:
Do we check that we don't have an empty field?

Reply:
```
Welcome to Node.js v18.15.0.
Type ".help" for more information.
> !!''
false
```

Reply:
We'll use the first non-blank line. It's of course not perfect.